api/proxy/credentials.go

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package proxy
+
+import (
+	"errors"
+
+	"github.com/hashicorp/boundary/api/targets"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	usernamePasswordCredentialType = "username_password"
+	sshPrivateKeyCredentialType    = "ssh_private_key"
+)
+
+// UsernamePassword contains username and password credentials
+type UsernamePassword struct {
+	Username string `mapstructure:"username"`
+	Password string `mapstructure:"password"`
+
+	Raw *targets.SessionCredential
+	// Consumed can be set by the caller to indicate that the credential has
+	// been used, e.g. displayed to the user
+	Consumed bool
+}
+
+// SshPrivateKey contains the username and private key with optional passphrase
+// for the key
+type SshPrivateKey struct {
+	Username   string `mapstructure:"username"`
+	PrivateKey string `mapstructure:"private_key"`
+	Passphrase string `mapstructure:"private_key_passphrase"`
+
+	Raw *targets.SessionCredential
+	// Consumed can be set by the caller to indicate that the credential has
+	// been used, e.g. displayed to the user
+	Consumed bool
+}
+
+type Credentials struct {
+	UsernamePassword []UsernamePassword
+	SshPrivateKey    []SshPrivateKey
+	// Unspecified are credentials that do not match one of the types above
+	Unspecified []*targets.SessionCredential
+}
+
+func (c Credentials) UnconsumedSessionCredentials() []*targets.SessionCredential {
+	out := make([]*targets.SessionCredential, 0, len(c.SshPrivateKey)+len(c.UsernamePassword)+len(c.Unspecified))
+
+	// Unspecified credentials cannot be consumed
+	out = append(out, c.Unspecified...)
+
+	for _, c := range c.SshPrivateKey {
+		if !c.Consumed {
+			out = append(out, c.Raw)
+		}
```

Comment:
No, because it's not consumed. The way the code works in the old location is that it's up to the consumer to mark when it's consumed; if we mark it consumed here then if various consumers want to check to see if they know how to consume a given credential it will already be marked.